pkg/cri/resource-manager/control/blockio/blockio.go

```diff
@@ -133,25 +164,282 @@ func (ctl *blockio) BlockIOClass(c cache.Container) string {
+			if err != nil {
+				return oci, err
+			}
+			for devNode, blockDeviceInfo := range blockDevices {
+				if weight != -1 || leafWeight != -1 {
+					if ios, found := currentIOSchedulers[devNode]; found {
+						if ios != "bfq" && ios != "cfq" {
+							log.Warn("weight/leaf_weight has no effect on device %#v due to "+
+								"incompatible io-scheduler %#v (bfq or cfq required)", devNode, ios)
+						}
+					}
+					oci.WeightDevice = append(oci.WeightDevice,
+						cgroups.OciWeightDeviceParameters{
+							Major:      blockDeviceInfo.Major,
+							Minor:      blockDeviceInfo.Minor,
+							Weight:     weight,
+							LeafWeight: leafWeight})
+				}
+				if throttleReadBps != -1 {
+					oci.ThrottleReadBpsDevice = append(oci.ThrottleReadBpsDevice,
+						cgroups.OciRateDeviceParameters{
+							Major: blockDeviceInfo.Major,
+							Minor: blockDeviceInfo.Minor,
+							Rate:  throttleReadBps})
+				}
+				if throttleWriteBps != -1 {
+					oci.ThrottleWriteBpsDevice = append(oci.ThrottleWriteBpsDevice,
+						cgroups.OciRateDeviceParameters{
+							Major: blockDeviceInfo.Major,
+							Minor: blockDeviceInfo.Minor,
+							Rate:  throttleWriteBps})
+				}
+				if throttleReadIOPS != -1 {
+					oci.ThrottleReadIOPSDevice = append(oci.ThrottleReadIOPSDevice,
+						cgroups.OciRateDeviceParameters{
+							Major: blockDeviceInfo.Major,
+							Minor: blockDeviceInfo.Minor,
+							Rate:  throttleReadIOPS})
+				}
+				if throttleWriteIOPS != -1 {
+					oci.ThrottleWriteIOPSDevice = append(oci.ThrottleWriteIOPSDevice,
+						cgroups.OciRateDeviceParameters{
+							Major: blockDeviceInfo.Major,
+							Minor: blockDeviceInfo.Minor,
+							Rate:  throttleWriteIOPS})
+				}
+			}
+		}
+		err = utils.JoinErrors(errors)
+		if err != nil {
+			return oci, err
+		}
+	}
+	return oci, nil
+}
+
+func parseAndValidateInt64(fieldName string, fieldContent string,
+	defaultValue int64, min int64, max int64) (int64, error) {
+	// Returns field content
+	if fieldContent == "" {
+		return defaultValue, nil
+	}
+	qty, err := resource.ParseQuantity(fieldContent)
+	if err != nil {
+		return defaultValue, fmt.Errorf("syntax error in %#v (%#v)", fieldName, fieldContent)
+	}
+	value := qty.Value()
+	if min != -1 && min > value {
+		return defaultValue, fmt.Errorf("value of %#v (%#v) smaller than minimum (%#v)", fieldName, value, min)
+	}
+	if max != -1 && value > max {
+		return defaultValue, fmt.Errorf("value of %#v (%#v) bigger than maximum (%#v)", fieldName, value, max)
+	}
+	return value, nil
+}
+
+type BlockDeviceInfo struct {
+	Major  int64
+	Minor  int64
+	Origin string // for traceability: why this device is configured
+}
+
+func configurableBlockDevices(devWildcards []string) (map[string]BlockDeviceInfo, error) {
+	// Return map {devNode: BlockDeviceInfo}
+	// Example: {"/dev/sda": {Major:8, Minor:0, Origin:"from symlink /dev/disk/by-id/ata-VendorXSSD from wildcard /dev/disk/by-id/*SSD*"}}
+	errors := []error{}
```

Comment:
nitpick: why do we collect these errors if we never do anything with them beyond collecting ?

Reply:
Great finding. :) Signs of evolution already in the first commit... I'll recheck what needs to be done.